### ISSUE TITLE
[5.0.0] Web building slow

### ISSUE THREAD

**Issue by jniemin** [storybookjs/storybook#5935]
**Describe the bug**
I have an app build with create-react-app. Upgrading @storybook/react from 4.1.13 to 5.0.0 increased webpack build time from 500ms to 5000ms. Initial startup was increased from 8s to 11s, but the bigger issues is that while doing small change to code completion time went 10 fold.

**To Reproduce**
1. Start storybook
1. Make modification to existing code
1. See webpack build time from command line

**Expected behavior**
No increase built time

**System:**
 - OS: MacOS
 - Framework: React 16.8.4, Create-react-app 2.1.5
 - Addons: No addons

**Comment by shilman** [comment]
Any chance you can provide a repro repo?

**Comment by jniemin** [comment]
Unfortunately it's private repo. I'll see if I can create custom repo that reproduces the issue

**Comment by jniemin** [comment]
I created a new create-react-app and added storybook as dependency. With quick test I saw some 20-30ms longer built times with 5.0.0. Not sure if the difference grows bigger when project gets bigger.

**Comment by jniemin** [comment]
I finally had time to dig bit deeper on this. Seems that it's all about our create-react-app and typescript that causes the slow compilation times. Only thing I wonder is that why it increased dramatically between v4 and v5 of storybook. Code and build configuration was same on both versions

**Comment by shilman** [comment]
@jniemin Thanks for taking a look at this. I've seen other issues mentioning performance issues with typescript. I'm wondering whether this problem got worse due to changes we made in 5.0 that changed the way webpack is configured (see https://github.com/storybooks/storybook/issues/6081 for an umbrella issue)

**Comment by jniemin** [comment]
It seemed to got worse upgrading from 4->5, since build time was increase about 5 second in our system and only change we had was upgrading storybook. Our project is medium to large sized.

To fix typescript slowness I upgraded create-react-app to 2.1.8 (https://github.com/facebook/create-react-app/issues/5820) which enabled async type checking by default and also set in our tsconfig `"skipLibCheck": true`. Now we are back at where we were with 4.1.13 version.

**Comment by Andersos** [comment]
I´m also seeing slower build times but Im not using Typescript.
This is the repo and it is now on storybook 5 https://github.com/getas/styleguide 

The version Im comparing to running Storybook 4 only changing these dependencies 
`yarn add @storybook/addons@4.0.0 @storybook/addon-links@4.0.0 @storybook/addon-knobs@4.0.0 @storybook/addon-info@4.0.0 @storybook/addon-actions@4.0.0 @storybook/react@4.0.9`


$ yarn build-storybook (version 4)
✨  Done in 19.00s.
✨  Done in 5.38s.
✨  Done in 5.62s.

$ yarn build-storybook (verison 5)
✨  Done in 32.10s.
✨  Done in 7.80s.
✨  Done in 7.69s.

It looks like there are improvements from first time to the second and third time running the script after eachother. But also it looks to me that the build times have increased.

**Comment by steffenmllr** [comment]
Same here we went from 40sek to over 2 minutes for the initial build and 2-15sek for the rebuild. Thinking about switching back to 4.x. I'm using babel with typescript.

**Comment by shilman** [comment]
@steffenmllr @jniemin Anybody who's using typescript, please take a look at this issue for a fix: https://github.com/storybooks/storybook/issues/6430

TLDR; you should be using `react-docgen-typescript-loader` instead of `react-docgen-typescript-webpack-plugin`

**Comment by steffenmllr** [comment]
thanks for the hint but I'm not using `react-docgen-typescript` or `react-docgen-typescript-webpack-plugin`.

**Comment by shilman** [comment]
Ugh, sorry to hear that. @ndelangen any performance hints for @steffenmllr ?

**Comment by shilman** [comment]
For anybody who is interested in performance and doing any webpack customization, it's important to understand the webpack configuration. I added a `--debug-webpack` flag to print out the manager & preview webpack configs. Please take a look and see what's being applied in 5.x vs 4.x and this will most likely explain the performance differences.

If it's getting slower without any webpack customization, please let us know!

**Comment by jniemin** [comment]
@shilman Like I wrote in my previous comment, this helped for our project to get back to decent build times.

> To fix typescript slowness I upgraded create-react-app to 2.1.8 (facebook/create-react-app#5820) which enabled async type checking by default and also set in our tsconfig "skipLibCheck": true.

**Comment by nonameolsson** [comment]
I can agree that it is very slow. I haven't used Storybook v4, only the latest one. 
![image](https://user-images.githubusercontent.com/1493018/58489492-ac143500-816b-11e9-94c2-720d641108bb.png)

**Comment by shilman** [comment]
@nonameolsson are you using typescript? did you use the workaround here?

**Comment by KamalAman** [comment]
I am using native javascript and finding the same issue. I did a performance profile and I see the the culprit is `storybook-addon-react-docgen` which uses `react-docgen` underneath.

It seems like `framework-preset-react-docgen.js` adds the plugin into the bundle process. If I add an early return to framework-preset file (and not use docgen) then my storybook build time drops from 4 minutes to 40 seconds. 

It seems like docgen is running on files that are not `jsx` including all file in node_modules. 

![image](https://user-images.githubusercontent.com/2692357/58829695-7ad9b000-8616-11e9-92a2-d6a5a3949105.png)

Close up inspection of visitingChildren
![image](https://user-images.githubusercontent.com/2692357/58829769-a8bef480-8616-11e9-938c-8b39d091bce3.png)

In addition, I don't understand why docgen is needed, I am not using the prop addon for storybook. Can we skip adding it by default or provide a way to disable it?

**Comment by hjnilsson** [comment]
@KamalAman Is there any way to add this early return without modifying storybook itself? The build times are crippling use of storybook in our project.

**Comment by shilman** [comment]
@KamalAman Thanks for the detailed report. We've recently started shipping presets to make it easy to configure storybook for common use cases (e.g. add typescript or scss). I think docgen could be good candidate for that for 6.0 (this will be a breaking change). I'm also thinking maybe the current default is not configured properly if it's also running on node_modules and all JS files... cc @igor-dv

**Comment by shilman** [comment]
On a related note @jessepinho and @leoyli also went through the perf debugging process for typescript and found that things sped up dramatically when he used babel to handle typescript instead of `awesome-ts-loader`. Referencing them so they can add some color here.

Also cc @mrmckeb who is working on turning the CRA config into a preset. Maybe we can make the docgen stuff opt-out in the process somehow? (And maybe opt-in in 6.0?!)

**Comment by mrmckeb** [comment]
Yes, so the CRA preset doesn't include doc-gen at all, and I thought about that yesterday and will make it opt-in I think.

So the preset would include doc-gen for JS and TS, and you just need to turn it on. Sound fair?

**Comment by shilman** [comment]
I'm not sure about opt-out vs opt-in but we can start with opt-in. Seems like ts loading is also a pain point. Any thoughts on that?

**Comment by mrmckeb** [comment]
@shilman, it shouldn't be... we are using TS with React on several projects in my company and don't have performance issues - as CRA is already using the `babel-loader`. So the preset I'm working on should be immune to these issues!

**Comment by KamalAman** [comment]
@hjnilsson 	I added the following as a pre `storybook` npm script hook. It adds an early return to the docgen present :(

`prestorybook: node ./make-storybook-faster.js`

make-stroybook-faster.js

```
const replace = require('replace-in-file');

replace.sync({
  files: 'node_modules/@storybook/react/dist/server/framework-preset-react-docgen.js',
  from: /^function babel\(config\) \{.*/m,
  to: 'function babel(config) { return config;',
});

```

Building storybook on the ci server went from 10 minutes to ~2minutes

**Comment by macrozone** [comment]
> @hjnilsson I added the following as a pre `storybook` npm script hook. It adds an early return to the docgen present :(
> 
> `prestorybook: node ./make-storybook-faster.js`
> 
> make-stroybook-faster.js
> 
> ```js
> const replace = require('replace-in-file');
> 
> replace.sync({
>   files: 'node_modules/@storybook/react/dist/server/framework-preset-react-docgen.js',
>   from: /^function babel\(config\) \{.*/m,
>   to: 'function babel(config) { return config;',
> });
> ```
> 
> Building storybook my ci server went from 10 min to ~2min
> 
> It still modifies storybook, but in an automated way.
> No Guarantees!

did not work for me, rebuilding on changed file still takes > 10seconds, which is too long for hot module reload

**Comment by bote795** [comment]
> I am using native javascript and finding the same issue. I did a performance profile and I see the the culprit is `storybook-addon-react-docgen` which uses `react-docgen` underneath.
> 
> It seems like `framework-preset-react-docgen.js` adds the plugin into the bundle process. If I add an early return to framework-preset file (and not use docgen) then my storybook build time drops from 4 minutes to 40 seconds.
> 
> It seems like docgen is running on files that are not `jsx` including all file in node_modules.
> 
> ![image](https://user-images.githubusercontent.com/2692357/58829695-7ad9b000-8616-11e9-92a2-d6a5a3949105.png)
> 
> Close up inspection of visitingChildren
> ![image](https://user-images.githubusercontent.com/2692357/58829769-a8bef480-8616-11e9-938c-8b39d091bce3.png)
> 
> In addition, I don't understand why docgen is needed, I am not using the prop addon for storybook. Can we skip adding it by default or provide a way to disable it?

sorry, but how were you able to get the profiling setup, I am building with `build-storybook` but I can't pass in `--inspect-brk` to be able to attach the browser to do a profile (sorry if this is off topic) my build takes a while and trying to debug why

**Comment by KamalAman** [comment]
@bote795 execute:

`node --inspect ./node_modules/@storybook/react/bin/index.js`

Open chrome to `chrome://inspect` and debug from there

**Comment by matheo** [comment]
@shilman I perceive a speed improvement in the build time of the latest `6.0.0-apha`, but I wonder if there will be improvements loading huge projects. Currently, the size of my main bundle is 25Mb, with vendors it's up to 32Mb, so it would be good to have some kind of lazy-loading of the Stories per Category, so we can improve the UX and loading times.

Thanks in advance!

**Comment by shilman** [comment]
@matheo yeah we're working on a variety of related projects. I don't have numbers yet but 6.0 will include some massive perf improvements for large storybooks.

**Comment by KamalAman** [comment]
I'm am definitely seeing more stable start-storybook build and decreased build time in `6.0.0-beta.0`. 

The places where I am noticing slow downs are: 
- initial load of the storybook site, it is in the loading state for 20ish seconds.
- build-storybook crashes to due running out of memory when minifying the code which did not happen in v5 [6408](https://github.com/storybookjs/storybook/issues/6408)

**Comment by jayair** [comment]
Hi all, I recently upgraded from 5 to 6.0.13 and Storybook is unusably slow.

I'm using Safari. A large number of stories are using the `storiesOf` API.

Here are some of the problems I'm noticing:

- The initial load takes 22s
- Making a small change causes the browser to freeze for at least 5s
- After a few minutes, Safari shows the notice that the browser window is taking too much memory and the page starts to become unresponsive.

Is there something I can do to debug what's going on?

Thanks.

**Comment by Obiwarn** [comment]
@shliman FYI: Storybook 6 got so slow we tend to use it less frequently.



  Storybook 6.1.0-alpha.17 started                  │
│   2.03 min for manager and 3.85 min for preview  


^^^ This is with:
```
"typescript": {
    "reactDocgen": "none"
}
```

**Comment by torkelrogstad** [comment]
Where am I supposed to put the `reactDocgen: "none"` snippet?

**Comment by Obiwarn** [comment]
@torkelrogstad 

/.storybook/main.js

**Comment by shilman** [comment]
@Obiwarn we're actively working on perf now. we've got a round of improvements in 6.1 and will continue throughout the 6.x cycle. this is our top priority now.

**Comment by Zaynex** [comment]
> @shilman I'm in 6.0.16 having a huge slowdown as well. You, in another issue (I can't seem to find it) suggested to turn off reactDocgen entirey and it worked, everything started building in a blink of an eye... but now I don't have controls anymore.
> 
> If I let reactDoggen turned on, it normally hangs at `70% sealing React Docgen Typescript Plugin`.
> 
> What you suggested:
> 
> ```
> "typescript": {
>     "reactDocgen": "none"
> }
> ```
> 
> So, or I'll work with the slowdown + controls or fast without controls?
> 
> (If you want to, I can open a new issue because of the version difference)

I want know if there any config that can ignore some npm package when run  React Docgen Typescript Plugin.

This wasn't a serious problem at first, until I import `babylon.js`（3d library).

**Comment by shilman** [comment]
@Zaynex 

- Try upgrading to 6.1. There were some docgen improvements there
- Try setting `reactDocgen` to 'react-docgen'

You can try playing around with the include paths of 'react-docgen-plugin', but I haven't seen any huge speedups from that.

**Comment by Zaynex** [comment]
I'm using `@storybook/cli:^6.0.21` .
After using `reactDocgen to 'react-docgen'`, webpack hot reload time from 20000ms - 50000ms  to 1000ms - 3000ms.
@shilman  Thanks, man.

**Comment by bregenspan** [comment]
In addition to the the change of configuring `typescript.reactDocgen: 'react-docgen'`, something that gave a huge performance boost for my company's Storybook was updating the glob patterns included in our `stories` config.

We had started with:

```
// main.js

stories: ['../**/*.stories.@(js|jsx|ts|tsx|mdx)'],
```

and we found that updating to a more specific glob pattern shaved off ~75% of what was previously a ~20 second build time:

```
const path = require('path');
/* ... */
stories: [path.join(__dirname, '..', 'src/**/*.stories.@(js|jsx|ts|tsx|mdx')],
```

(I still want to dig a little more into exactly why we get so wildly different a result between these two configurations, but posting here in case helpful to anyone else seeing slow performance).

**Comment by shilman** [comment]
@bregenspan the first glob includes `node_modules` 😭

**Comment by sandwichsudo** [comment]
> #### .storybook/main.js
> ```
> const path = require('path');
> /* ... */
> stories: [path.join(__dirname, '..', 'src/**/*.stories.@(js|jsx|ts|tsx|mdx')],
> ```

This was a massive boost for us, thanks so much for sharing

**Comment by koba04** [comment]
I have a similar issue with storybook v6 on `react-transition-group` and it is a blocker to update the storybook version.

```
    __child-HtmlWebpackPlugin_0  6.32 KiB  HtmlWebpackPlugin_0  HtmlWebpackPlugin_0
    Entrypoint HtmlWebpackPlugin_0 = __child-HtmlWebpackPlugin_0
    [./node_modules/html-webpack-plugin/lib/loader.js!./node_modules/@storybook/core-common/dist/cjs/templates/index.ejs] 2 KiB {HtmlWebpackPlugin_0} [built]
ℹ ｢wdm｣: Compiled successfully.  // Takes about 3 minutes to display the following message
╭─────────────────────────────────────────────────────╮
│                                                     │
│   Storybook 6.2.9 started                           │
│   15 s for manager and 14 s for preview             │
│                                                     │
│    Local:            http://localhost:6006/         │
│    On your network:  http://192.168.116.80:6006/    │
│                                                     │
╰─────────────────────────────────────────────────────╯

```

You can reproduce it on https://github.com/reactjs/react-transition-group/pull/741.

**Comment by levymetal** [comment]
I'm upgrading an enterprise project from Webpack 4 to Webpack 5. I'm seeing significantly slower rebuilds in Storybook on Webpack 5 (~800ms to 5500ms). 

`reactDocgen: 'none'` shaves this down to 3500ms. Removing all css brings it down to 1500ms.

There are numerous issues at play and I'm not sure if any are related to Storybook.

For the css, I believe the culprit to be `css-loader` (unless I'm missing some really obvious config). I've created a minimal repro (https://github.com/levymetal/webpack5-css-loader-repro) and posted it on an existing Webpack issue: https://github.com/webpack/webpack/issues/12102#issuecomment-909968808.

I'm still trying to figure out why the rebuild is 2x slower with `reactDocgen: 'none'` and no css and if this can be isolated to Storybook or Webpack.

**Comment by shilman** [comment]
@levymetal which version of storybook are you using? i think there's a `react-docgen-typescript` perf regression in `6.4-alpha`, but it should be OK in `6.3.7`. would be great if you could verify that in your project. 🙏 

this is the first i'm hearing about the `css-loader` issue though 🤦

**Comment by levymetal** [comment]
@shilman I'm on `6.3.7`, I'm pretty sure the 2000ms saving by using `reactDocgen: 'none'` is on `6.3.7`, but I can't say for sure, will need to test again. I haven't tested that part in a while because `reactDocgen: 'none'` is fine for us.

I was wrong about `css-loader`, the issue in my repro was actually caused by `mini-css-extract-plugin`. Replacing that with `style-loader` or using `experimentalUseImportModule: true` fixes the repro I posted above, and may work for others, but sadly doesn't fix the performance issues in the real repo I'm working on. So there's still something else going on. I'm back to the drawing board, but slowly but surely narrowing down the culprits.

**Comment by levymetal** [comment]
I've finally been able to make a repro of the issue we're experiencing. It seems to have something to do with sass imports. When using `@import` in css modules, all css is recompiled on every rebuild, but only when running the Storybook server. The webpack dev server is fine. 

I've updated the repro here: https://github.com/levymetal/webpack5-css-loader-repro

The readme contains detailed instructions, but here's the TLDR: 

 - run `yarn generate` (generates 300 dummy components)
 - run `yarn start` to run the webpack dev server. Take note of the rebuild time (no issues)
 - run `yarn storybook`. Note rebuilds are significantly slower (all css is recompiled)

If you repeat these steps using webpack 4 (switch to the [`webpack4`](https://github.com/ferocia/johanna-platform/pull/8912/files) branch), there's no difference between the Storybook server and webpack dev server; rebuilds are fast in both.

**Comment by valentinpalkovic** [comment]
Hey @levymetal,

Do you still encounter the same issue with the newest version of Storybook?

I have updated the `package.json` of your repro like this:

```
{
  "name": "webpack5-css-loader-repro",
  "version": "0.0.0",
  "description": "compare build perf between webpack 4/5",
  "main": "index.js",
  "private": true,
  "scripts": {
    "generate": "./script/generate_components",
    "start": "yarn && yarn webpack serve --open",
    "storybook": "yarn && start-storybook -p 9009"
  },
  "author": "",
  "license": "",
  "dependencies": {
    "@babel/core": "^7.20.5",
    "@babel/plugin-transform-runtime": "^7.19.6",
    "@babel/preset-env": "^7.20.2",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.18.6",
    "@babel/register": "^7.18.9",
    "acorn": "^8.4.1",
    "babel-loader": "^9.1.0",
    "core-js": "^3.26.1",
    "css-loader": "^6.7.2",
    "html-entities": "^2.3.3",
    "sass": "^1.56.1",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "sass-loader": "^12.2.0",
    "style-loader": "^3.3.1",
    "thread-loader": "^3.0.4",
    "typescript": "^4.3.5",
    "webpack": "^5.75.0",
    "webpack-cli": "^5.0.1",
    "webpack-dev-server": "^4.11.1"
  },
  "devDependencies": {
    "@storybook/builder-webpack5": "^6.5.14",
    "@storybook/manager-webpack5": "^6.5.14",
    "@storybook/react": "^6.5.14",
    "@types/react-dom": "^17.0.9",
    "prettier": "^2.3.2"
  }
}
```

- I have updated all build and Storybook related dependencies
- I have deleted `node-sass` and installed `sass` (Dart Sass) instead, which is the [official recommendation](https://github.com/webpack-contrib/sass-loader#getting-started).
- Starting Storybook and changing the CSS of one component will almost immediately show me the changes in Storybook

Please reopen this issue, if you think it is still relevant
